File: Baekjoon/Class2/32.10828 스택.py
import sys
input = sys.stdin.readline
write = sys.stdout.write

class Stack: #연결 리스트 스택
    class Node: #연결 리스트 노드
        def __init__(self, data, next):
            self.data = data
            self.next = next
    def __init__(self):
        self._top = None #첫 노드
        self._size = 0
    def push(self, data):
        self._size += 1
        node_to_push = self.Node(data, self._top)
        self._top = node_to_push
    def pop(self):
        if self.empty(): raise IndexError("stack is empty")
        self._size -= 1
        data = self._top.data
        self._top = self._top.next
        return data
    def size(self):
        return self._size
    def empty(self):
        return self._size == 0
    def top(self):
        if self.empty(): raise IndexError("stack is empty")
        return self._top.data

def main():
    n = int(input())
    nums = Stack()

    for _ in range(n):
        line = input().split()
        command = line[0]
        try:
            match(command):
                case "push":
                    nums.push(int(line[1]))
                case "pop":
                    write(f"{nums.pop()}\n")
                case "size":
                    write(f"{nums.size()}\n")
                case "empty":
                    write(f"{int(nums.empty())}\n")
                case "top":
                    write(f"{nums.top()}\n")
        except IndexError: #스택 빔
            write("-1\n")

main()

''' 라이브러리 사용
from collections import deque
import sys
input = sys.stdin.readline
write = sys.stdout.write

n = int(input())
nums = deque()

for _ in range(n):
    line = input().split()
    command = line[0]
    match(command):
        case "push":
            nums.append(int(line[1]))
        case "pop":
            if len(nums) == 0:
                write("-1\n")
                continue
            write(f"{nums.pop()}\n")
        case "size":
            write(f"{len(nums)}\n")
        case "empty":
            write(f"{int(not nums)}\n")
        case "top":
            if not nums:
                write("-1\n")
                continue
            write(f"{nums[-1]}\n")
'''

File: Baekjoon/Class2/32.10828 스택.js
const inputFile = process.platform === "linux" ? "/dev/stdin" : "./input.txt";
const input = require("fs").readFileSync(inputFile, "utf8").trim().split("\n");

class Stack { // 연결 리스트 스택
    static Node = class {
        constructor(data, next) { // 연결 리스트 노드
            this.data = data;
            this.next = next;
        }
    }
    #top; // 첫 노드
    #size;
    constructor() {
        this.#top = null;
        this.#size = 0;
    }
    push(data) {
        this.#size++;
        const nodeToPush = new Stack.Node(data, this.#top);
        this.#top = nodeToPush;
    }
    pop() {
        if (this.empty) return undefined;
        this.#size--;
        const data = this.#top.data;
        this.#top = this.#top.next;
        return data;
    }
    get size() {
        return this.#size;
    }
    get empty() {
        return this.#size === 0;
    }
    get top() {
        if (this.empty) return undefined;
        return this.#top.data;
    }
}

const n = Number(input[0]);
const nums = new Stack();
const out = [];

for(let i = 1; i <= n; i++) {
    const line = input[i].split(" ");
    const command = line[0].trim();
    let num;
    switch(command) {
        case "push":
            nums.push(Number(line[1]));
            break;
        case "pop":
            num = nums.pop();
            if(num === undefined) out.push("-1");
            else out.push(num);
            break;
        case "size":
            out.push(nums.size);
            break;
        case "empty":
            out.push(nums.empty ? 1 : 0);
            break;
        case "top":
            num = nums.top;
            if(num === undefined) out.push("-1");
            else out.push(num);
            break;
    }
}

console.log(out.join("\n"));

/* 배열 사용
const inputFile = process.platform === "linux" ? "/dev/stdin" : "./input.txt";
const input = require("fs").readFileSync(inputFile, "utf8").trim().split("\n");

const n = Number(input[0]);
const nums = [];
const out = [];

for(let i = 1; i <= n; i++) {
    const line = input[i].split(" ");
    const command = line[0].trim();
    let num;
    switch(command) {
        case "push":
            nums.push(Number(line[1]));
            break;
        case "pop":
            out.push(nums.length ? nums.pop() : -1);
            break;
        case "size":
            out.push(nums.length);
            break;
        case "empty":
            out.push(nums.length ? 0 : 1);
            break;
        case "top":
            out.push(nums.length ? nums[nums.length - 1] : -1);
            break;
    }
}

console.log(out.join("\n"));
*/